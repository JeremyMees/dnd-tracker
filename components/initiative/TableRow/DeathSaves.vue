<script setup lang="ts">
import { useToast } from '~/components/ui/toast/use-toast'

const props = defineProps<{
  item: InitiativeSheetRow
  sheet: InitiativeSheet | undefined
  update: (payload: Omit<Partial<InitiativeSheet>, NotUpdatable | 'campaign'>) => Promise<void>
}>()

const { toast } = useToast()
const { t } = useI18n()

const { checkDeathSaves } = deathSavesFunctions

function updateDeathSave(saveIndex: number, save: boolean): void {
  if (!props.sheet) return

  const index = getCurrentRowIndex(props.sheet, props.item.id)
  const rows = [...props.sheet.rows]

  if (index === -1) return
  if (!props.item.deathSaves) return

  const deathSaves = {
    save: [...props.item.deathSaves.save] as [boolean, boolean, boolean],
    fail: [...props.item.deathSaves.fail] as [boolean, boolean, boolean],
  }

  if (save) deathSaves.save[saveIndex] = !deathSaves.save[saveIndex]
  else deathSaves.fail[saveIndex] = !deathSaves.fail[saveIndex]

  rows[index] = {
    ...rows[index],
    deathSaves: {
      save: deathSaves.save,
      fail: deathSaves.fail,
    },
  }

  const { failed, saved } = checkDeathSaves(deathSaves)

  if ((saved && !failed && save) || (!saved && failed && !save)) {
    const toastType = saved ? 'stable' : 'died'

    toast({
      title: t(`components.initiativeTable.${toastType}.title`, { name: props.item.name }),
      description: t(`components.initiativeTable.${toastType}.textDeathSaves`, { name: props.item.name }),
      variant: toastType === 'stable' ? 'success' : 'destructive',
    })
  }

  props.update({ rows })
}
</script>

<template>
  <div
    class="grid gap-1 w-fit mx-auto"
    :class="{
      'bg-success/20 p-2 rounded-lg': item.deathSaves?.save.every(Boolean) && !item.deathSaves?.fail.every(Boolean),
      'bg-destructive/20 p-2 rounded-lg': item.deathSaves?.fail.every(Boolean) && !item.deathSaves?.save.every(Boolean),
    }"
  >
    <div
      v-for="(save, i) in [item.deathSaves?.save, item.deathSaves?.fail]"
      :key="`save-${i}`"
      class="grid grid-cols-3 gap-1 min-w-14"
    >
      <button
        v-for="(value, j) in save"
        :key="`${value}-${j}`"
        class="size-4 rounded border-2"
        :class="{
          'border-success bg-success/20': i === 0,
          'border-destructive bg-destructive/20': i === 1,
          '!bg-success': value && i === 0,
          '!bg-destructive': value && i === 1,
        }"
        @click="updateDeathSave(j, i === 0)"
      />
    </div>
  </div>
</template>
