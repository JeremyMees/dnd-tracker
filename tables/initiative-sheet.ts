import { createColumnHelper, type InitialTableState, type Row } from '@tanstack/vue-table'
import { actionsTable, expandButton, iconButton } from './generate-functions'

import {
  InitiativeTableRowName,
  InitiativeTableRowHealth,
  InitiativeTableRowActions,
  InitiativeTableRowConditions,
  InitiativeTableRowNotes,
  InitiativeTableRowInit,
  InitiativeTableRowAc,
  InitiativeTableRowDeathSaves,
  InitiativeTableRowConcentration,
  InitiativeTableRowModify,
} from '#components'

const columnHelper = createColumnHelper<InitiativeSheetRow>()

interface ColumnOptions {
  sheet: ComputedRef<InitiativeSheet | undefined>
  update: (payload: Omit<Partial<InitiativeSheet>, NotUpdatable>) => Promise<void>
  openQuickInitModal: () => void
}

export function generateColumns({ sheet, update, openQuickInitModal }: ColumnOptions) {
  const { t } = useI18n()

  return [
    columnHelper.display({
      enableGlobalFilter: false,
      id: 'index',
      cell: ({ row }) => h('div', { class: 'flex items-center gap-x-2 max-w-20' }, [
        h('span', { class: 'text-muted-foreground ml-1' }, row.index + 1),
        expandButton({
          content: t(`actions.${row.getIsExpanded() ? 'hide' : 'show'}`),
          expanded: row.getIsExpanded(),
          cb: () => row.toggleExpanded(),
        }),
      ]),
    }),
    columnHelper.accessor('name', {
      header: t('components.encounterTable.headers.name'),
      cell: ({ row }) => h(InitiativeTableRowName, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('initiative', {
      header: () => h('button', { class: 'flex items-center gap-x-2' }, [
        t('components.encounterTable.headers.init'),
        iconButton({
          icon: 'tabler:bolt-filled',
          content: t('components.encounterTable.quick'),
          color: 'warning',
          cb: () => openQuickInitModal(),
          disabled: !sheet.value?.rows.length,
          iconColor: 'text-warning',
        }),
      ]),
      cell: ({ row }) => h(InitiativeTableRowInit, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('health', {
      header: t('components.encounterTable.headers.health'),
      cell: ({ row }) => h(InitiativeTableRowHealth, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('ac', {
      header: t('components.encounterTable.headers.ac'),
      cell: ({ row }) => h(InitiativeTableRowAc, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.display({
      id: 'manage',
      header: t('components.encounterTable.headers.manage'),
      cell: ({ row }) => h(InitiativeTableRowActions, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('conditions', {
      header: t('components.encounterTable.headers.conditions'),
      cell: ({ row }) => h(InitiativeTableRowConditions, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('note', {
      header: t('components.encounterTable.headers.note'),
      cell: ({ row }) => h(InitiativeTableRowNotes, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('deathSaves', {
      header: t('components.encounterTable.headers.deathSaves'),
      cell: ({ row }) => h(InitiativeTableRowDeathSaves, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('concentration', {
      header: t('components.encounterTable.headers.concentration'),
      cell: ({ row }) => h(InitiativeTableRowConcentration, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.display({
      enableGlobalFilter: false,
      enableSorting: false,
      id: 'actions',
      cell: ({ row }) => h(InitiativeTableRowModify, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
  ]
}

export function expandedMarkup(row: Row<InitiativeSheetRow>) {
  return actionsTable(row.original)
}

export const initialState: InitialTableState = {}
