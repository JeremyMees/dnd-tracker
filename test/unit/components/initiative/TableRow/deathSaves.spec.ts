import { mountSuspended } from '@nuxt/test-utils/runtime'
import { describe, expect, it, vi, beforeEach } from 'vitest'
import DeathSaves from '~/components/initiative/TableRow/DeathSaves.vue'
import { sheet } from '~~/test/unit/fixtures/initiative-sheet'

interface Props {
  item: InitiativeSheetRow
  sheet: InitiativeSheet | undefined
  update: (payload: Omit<Partial<InitiativeSheet>, NotUpdatable | 'campaign'>) => Promise<void>
}

const mockUpdate = vi.fn()
const mockToast = vi.fn()

vi.mock('~/components/ui/toast/use-toast', () => ({
  useToast: () => ({
    toast: mockToast,
  }),
}))

const props: Props = {
  item: sheet.rows[0]!,
  sheet,
  update: mockUpdate,
}

describe('Initiative table row death saves', async () => {
  beforeEach(() => {
    mockUpdate.mockClear()
    mockToast.mockClear()
  })

  it('Should match snapshot', async () => {
    const component = await mountSuspended(DeathSaves, { props })

    expect(component.html()).toMatchSnapshot()
  })

  it('Should not render for lair type', async () => {
    const component = await mountSuspended(DeathSaves, {
      props: {
        ...props,
        item: { ...props.item, type: 'lair' },
      },
    })

    expect(component.find('[data-test-container]').exists()).toBeFalsy()
  })

  it('Should not render for summon type', async () => {
    const component = await mountSuspended(DeathSaves, {
      props: {
        ...props,
        item: { ...props.item, type: 'summon' },
      },
    })

    expect(component.find('[data-test-container]').exists()).toBeFalsy()
  })

  it('Should render for player type', async () => {
    const component = await mountSuspended(DeathSaves, {
      props: {
        ...props,
        item: {
          ...props.item,
          type: 'player',
          deathSaves: {
            save: [false, false, false],
            fail: [false, false, false],
          },
        },
      },
    })

    expect(component.find('[data-test-container]').exists()).toBeTruthy()
  })

  it('Should render 3 save and 3 fail buttons', async () => {
    const component = await mountSuspended(DeathSaves, { props })

    expect(component.findAll('[data-test-button="save"]').length).toBe(3)
    expect(component.findAll('[data-test-button="fail"]').length).toBe(3)
  })

  it('Should show success styling when all saves are successful', async () => {
    const component = await mountSuspended(DeathSaves, {
      props: {
        ...props,
        item: {
          ...props.item,
          type: 'player',
          deathSaves: {
            save: [true, true, true],
            fail: [false, false, false],
          },
        },
      },
    })

    expect(component.find('[data-test-container]').classes()).toContain('bg-success/20')
  })

  it('Should show destructive styling when all saves are failed', async () => {
    const component = await mountSuspended(DeathSaves, {
      props: {
        ...props,
        item: {
          ...props.item,
          type: 'player',
          deathSaves: {
            save: [false, false, false],
            fail: [true, true, true],
          },
        },
      },
    })

    expect(component.find('[data-test-container]').classes()).toContain('bg-destructive/20')
  })

  it('Should toggle save state when clicking a save button', async () => {
    const component = await mountSuspended(DeathSaves, {
      props: {
        ...props,
        item: {
          ...props.item,
          type: 'player',
          deathSaves: {
            save: [false, false, false],
            fail: [false, false, false],
          },
        },
      },
    })

    const buttons = component.findAll('[data-test-button="save"]')
    await buttons[0]!.trigger('click')

    expect(mockUpdate).toHaveBeenCalledWith({
      rows: expect.arrayContaining([
        expect.objectContaining({
          deathSaves: {
            save: [true, false, false],
            fail: [false, false, false],
          },
        }),
      ]),
    })
  })

  it('Should toggle fail state when clicking a fail button', async () => {
    const component = await mountSuspended(DeathSaves, {
      props: {
        ...props,
        item: {
          ...props.item,
          type: 'player',
          deathSaves: {
            save: [false, false, false],
            fail: [false, false, false],
          },
        },
      },
    })

    const buttons = component.findAll('[data-test-button="fail"]')
    await buttons[0]!.trigger('click')

    expect(mockUpdate).toHaveBeenCalledWith({
      rows: expect.arrayContaining([
        expect.objectContaining({
          deathSaves: {
            save: [false, false, false],
            fail: [true, false, false],
          },
        }),
      ]),
    })
  })

  it('Should not call update when sheet is undefined', async () => {
    const component = await mountSuspended(DeathSaves, {
      props: {
        ...props,
        sheet: undefined,
        item: {
          ...props.item,
          type: 'player',
          deathSaves: {
            save: [false, false, false],
            fail: [false, false, false],
          },
        },
      },
    })

    const buttons = component.findAll('[data-test-button]')
    await buttons[0]!.trigger('click')

    expect(mockUpdate).not.toHaveBeenCalled()
  })

  it('Should show stable toast when all saves are successful', async () => {
    const component = await mountSuspended(DeathSaves, {
      props: {
        ...props,
        item: {
          ...props.item,
          type: 'player',
          deathSaves: {
            save: [true, true, false],
            fail: [false, false, false],
          },
        },
      },
    })

    const buttons = component.findAll('[data-test-button="save"]')
    await buttons[2]!.trigger('click')

    expect(mockToast).toHaveBeenCalledWith({
      title: expect.stringMatching('components.initiativeTable.stable.title'),
      description: expect.stringMatching('components.initiativeTable.stable.textDeathSaves'),
      variant: 'success',
    })
  })

  it('Should show died toast when all fails are triggered', async () => {
    const component = await mountSuspended(DeathSaves, {
      props: {
        ...props,
        item: {
          ...props.item,
          type: 'player',
          deathSaves: {
            save: [false, false, false],
            fail: [true, true, false],
          },
        },
      },
    })

    const buttons = component.findAll('[data-test-button="fail"]')
    await buttons[2]!.trigger('click')

    expect(mockToast).toHaveBeenCalledWith({
      title: expect.stringMatching('components.initiativeTable.died.title'),
      description: expect.stringMatching('components.initiativeTable.died.textDeathSaves'),
      variant: 'destructive',
    })
  })

  it('Should not show toast when saves/fails are not complete', async () => {
    const component = await mountSuspended(DeathSaves, {
      props: {
        ...props,
        item: {
          ...props.item,
          type: 'player',
          deathSaves: {
            save: [true, false, false],
            fail: [false, false, false],
          },
        },
      },
    })

    const buttons = component.findAll('[data-test-button="save"]')
    await buttons[1]!.trigger('click')

    expect(mockToast).not.toHaveBeenCalled()
  })
})
