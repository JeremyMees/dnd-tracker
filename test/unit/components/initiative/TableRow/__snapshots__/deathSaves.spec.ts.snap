// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Initiative table row death saves > Should match snapshot 1`] = `
"<div data-test-container="" class="grid gap-1 w-fit mx-auto">
  <div class="grid grid-cols-3 gap-1 min-w-14"><button data-test-button="save" class="size-4 rounded border-2 border-success bg-success/20 !bg-success"></button><button data-test-button="save" class="size-4 rounded border-2 border-success bg-success/20"></button><button data-test-button="save" class="size-4 rounded border-2 border-success bg-success/20"></button></div>
  <div class="grid grid-cols-3 gap-1 min-w-14"><button data-test-button="fail" class="size-4 rounded border-2 border-destructive bg-destructive/20 !bg-destructive"></button><button data-test-button="fail" class="size-4 rounded border-2 border-destructive bg-destructive/20"></button><button data-test-button="fail" class="size-4 rounded border-2 border-destructive bg-destructive/20"></button></div>
</div>"
`;
